### PR TITLE
Add BlockResults gRPC service

### LINKED ISSUE #1095: ADR-101: Implement gRPC `BlockResultsService`
We need an implementation of the `BlockResultsService` from ADR-101 on the `feature/adr101-pull-companion` branch.

### PR DESCRIPTION
**Pull request by glnro**
Closes #1095 

This PR implements the gRPC service for BlockResults as specified in the ADR-101 Data Companion API
- `GetBlockResults` gRPC endpoint implemented
- Defined protos for BlockResults service
- Client implemented to interface with BlockResults `GetBlockResults` endpoint
- e2e Test
- Changelog entry
---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: proto/tendermint/services/block_results/v1/block_results.proto [hunk 1]
```diff
@@ -0,0 +1,23 @@
+syntax = "proto3";
+package tendermint.services.block_results.v1;
+
+import "tendermint/abci/types.proto";
+import "tendermint/types/params.proto";
+import "gogoproto/gogo.proto";
```
Comment (andynog) [inline proto/tendermint/services/block_results/v1/block_results.proto:6]: We should not use gogoproto, got this feedback from Thane

File: rpc/grpc/client/block_results_service.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+type blockResultServiceClient struct {
+	client v1.BlockResultsServiceClient
+}
+
+func (b blockResultServiceClient) GetBlockResults(ctx context.Context, req v1.GetBlockResultsRequest) (*ResultBlockResults, error) {
+	res, err := b.client.GetBlockResults(ctx, &v1.GetBlockResultsRequest{Height: req.Height})
+	if err != nil {
+		return &ResultBlockResults{}, err
```
Comment (andynog) [inline rpc/grpc/client/block_results_service.go:35]: It would be nice to have some logging and return a more friendly message to the client instead of bubbling up the error from the server. Please check this logic in the Block service client

File: rpc/grpc/server/server.go [hunk 3]
```diff
@@ -5,21 +5,29 @@ import (
 	"net"
 	"strings"
 
+	sm "github.com/cometbft/cometbft/state"
+	"github.com/cometbft/cometbft/store"
+
+	v1 "github.com/cometbft/cometbft/proto/tendermint/services/block_results/v1"
```
Comment (andynog) [inline rpc/grpc/server/server.go:11]: Better call this something else instead of 'v1' because there will be other services, maybe 'pbblkrsltsvc' or something like that

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	latestHeight := s.blockStore.Height()
+	if req.Height > latestHeight || req.Height < 0 {
+		return &v1.GetBlockResultsResponse{}, status.Error(codes.InvalidArgument, "Height is invalid.")
+	} else if req.Height == 0 {
+		height = latestHeight
+	}
+
+	res, err := s.stateStore.LoadFinalizeBlockResponse(height)
+	if err != nil {
+		return &v1.GetBlockResultsResponse{}, err
```
Comment (andynog) [inline rpc/grpc/server/services/blockresultservice/service.go:38]: As commented in the client logic this internal error might bubble up to the client, don't need to change here but just pointing it out based on comment in the client logic

File: test/e2e/tests/grpc_test.go [hunk 5]
```diff
@@ -30,3 +33,64 @@ func TestGRPC_Version(t *testing.T) {
...
+				res, err := gRPCClient.GetBlockResults(ctx, tc.request)
+				// First block tests
+				require.NoError(t, err)
+				require.NotNil(t, res)
+				require.Equal(t, res.Height, tc.expectedHeight)
+			}
+			for _, tc := range errorCases {
+				_, err = gRPCClient.GetBlockResults(ctx, tc.request)
+				require.Error(t, err)
+			}
```
Comment (andynog) [inline test/e2e/tests/grpc_test.go:93]: Maybe add a test for response to be nil ?

**Review by andynog — CHANGES_REQUESTED** [review]

File: test/e2e/tests/grpc_test.go [hunk 6]
```diff
@@ -30,3 +33,64 @@ func TestGRPC_Version(t *testing.T) {
...
+		gRPCClient, err := node.GRPCClient(ctx)
+		require.NoError(t, err)
+
+		for _, block := range blocks {
+			successCases := []struct {
+				expectedHeight int64
+				request        v1.GetBlockResultsRequest
+			}{
+				{first, v1.GetBlockResultsRequest{Height: first}},
+				{last, v1.GetBlockResultsRequest{}},
```
[the remark this answers is not in the document]
Comment (glnro) [reply test/e2e/tests/grpc_test.go:67]: In the endpoint definition it states that if no height is provided then the endpoint will fetch the latest block. The e2e test is indeed failing for the CI network setup (passes on simple) because of a race condition with the test setup. The test fails because the endpoint is fetching the latest endpoint which is greater than the value its being compared to (cached last block height).

File: proto/tendermint/services/block_results/v1/block_results.proto (hunk 1 above)
Comment (thanethomson) [reply proto/tendermint/services/block_results/v1/block_results.proto:6]: Yup. We should avoid it at all costs to better facilitate a future transition away from it, as per #731.

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+type blockResultsService struct {
+	stateStore sm.Store
+	blockStore *store.BlockStore
+	logger     log.Logger
+}
+
+// New creates a new CometBFT block results service server.
+func New(bs *store.BlockStore, ss sm.Store, logger log.Logger) brs.BlockResultsServiceServer {
+	return &blockResultsService{stateStore: ss, blockStore: bs, logger: logger}
```
Comment (thanethomson) [inline rpc/grpc/server/services/blockresultservice/service.go:24]:
```suggestion
	return &blockResultsService{
		stateStore: ss,
		blockStore: bs,
		logger: logger.With("service", "BlockResultsService"),
	}
```

Helps with logging context.

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// GetBlockResults returns the block results of the requested height.
+// If no height is given, the block results for the latest height are returned.
+func (s *blockResultsService) GetBlockResults(_ context.Context, req *brs.GetBlockResultsRequest) (*brs.GetBlockResultsResponse, error) {
+	logger := s.logger.With("endpoint", "GetBlockResults")
+	height := req.Height
+	latestHeight := s.blockStore.Height()
+	if req.Height > latestHeight || req.Height < 0 {
+		logger.Error("Failed to validate height from BlockResults request", "height", height)
```
Comment (thanethomson) [inline rpc/grpc/server/services/blockresultservice/service.go:34]:
```suggestion
```

No need for error logs in "bad request"-type responses. This might flood operators.

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		return nil, status.Error(codes.InvalidArgument, "Invalid request: height must be between 0 and the last effective height.")
```
Comment (thanethomson) [inline rpc/grpc/server/services/blockresultservice/service.go:35]:
```suggestion
		return nil, status.Errorf(codes.InvalidArgument, "Height must be between 0 and the last effective height (%d)", latestHeight)
```

Might be useful to instruct the caller as to what the latest height is at the time of the request. Also, the `Invalid request` part of the message isn't really necessary given the error code we're sending back.

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	} else if req.Height == 0 {
+		height = latestHeight
+	}
+
+	res, err := s.stateStore.LoadFinalizeBlockResponse(height)
+	if err != nil {
+		logger.Error("Error fetching BlockResults", "height", height, "err", err)
+		return nil, status.Errorf(codes.Internal, "Internal server error: cannot fetch BlockResults for height %d", height)
```
Comment (thanethomson) [inline rpc/grpc/server/services/blockresultservice/service.go:43]:
```suggestion
		return nil, status.Error(codes.Internal, "Internal server error")
```

Since the error and the height are being logged, operators can dig into what's going on here. As a stretch goal, I'd recommend looking into using this request tracing approach I've implemented here: https://github.com/cometbft/cometbft/pull/1142/files#diff-2cdfb60e6f68e99a22b64a5d79cb26e00adb76e670b0f6d144b7912bf1ae2354R84

That way, if a user asks the operator about why they're getting internal server errors, the operator will have logs to link their specific requests to.

**Review by thanethomson — COMMENTED** [review]
Seems like the E2E test's failing at the moment?

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+	}
+}
+
+// GetBlockResults returns the block results of the requested height.
+// If no height is given, the block results for the latest height are returned.
+func (s *blockResultsService) GetBlockResults(_ context.Context, req *brs.GetBlockResultsRequest) (*brs.GetBlockResultsResponse, error) {
+	logger := s.logger.With("endpoint", "GetBlockResults")
+	height := req.Height
+	latestHeight := s.blockStore.Height()
+	if req.Height > latestHeight || req.Height < 0 {
```
Comment (jmalicevic) [inline rpc/grpc/server/services/blockresultservice/service.go:37]: As we now support separate pruning of the blockstore and the abci results , it can happen that we don't have blocks but we do have the abci results. What would make more sense maybe is try and fetch the results and if there is no error return that.

File: rpc/grpc/client/block_results_service.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+	TxsResults            []*abci.ExecTxResult      `json:"txs_results"`
+	FinalizeBlockEvents   []*abci.Event             `json:"finalize_block_events"`
+	ValidatorUpdates      []*abci.ValidatorUpdate   `json:"validator_updates"`
+	ConsensusParamUpdates *cmtproto.ConsensusParams `json:"consensus_param_updates"`
+	AppHash               []byte                    `json:"app_hash"`
+}
+
+// BlockResultsServiceClient provides the block results of a given height (or latest if none provided).
+type BlockResultsServiceClient interface {
+	GetBlockResults(ctx context.Context, req brs.GetBlockResultsRequest) (*ResultBlockResults, error)
```
Comment (thanethomson) [inline rpc/grpc/client/block_results_service.go:26]: This should just be:
```suggestion
	GetBlockResults(ctx context.Context, height int64) (*ResultBlockResults, error)
```

File: rpc/grpc/client/block_results_service.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+	abci "github.com/cometbft/cometbft/abci/types"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
+
+	"github.com/cosmos/gogoproto/grpc"
+
+	brs "github.com/cometbft/cometbft/proto/tendermint/services/block_results/v1"
+)
+
+type ResultBlockResults struct {
```
Comment (thanethomson) [inline rpc/grpc/client/block_results_service.go:15]: Does it not make sense to call this something like `BlockResults`? Seems simpler.

File: rpc/grpc/server/server.go [hunk 14]
```diff
@@ -5,21 +5,29 @@ import (
...
 type Option func(*serverBuilder)
 
 type serverBuilder struct {
-	listener       net.Listener
-	versionService pbversionsvc.VersionServiceServer
-	logger         log.Logger
-	grpcOpts       []grpc.ServerOption
+	listener            net.Listener
+	versionService      pbversionsvc.VersionServiceServer
+	blockresultsService brs.BlockResultsServiceServer
```
Comment (thanethomson) [inline rpc/grpc/server/server.go:28]: `blockResultsService`?

File: rpc/grpc/client/block_results_service.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+	TxsResults            []*abci.ExecTxResult      `json:"txs_results"`
+	FinalizeBlockEvents   []*abci.Event             `json:"finalize_block_events"`
+	ValidatorUpdates      []*abci.ValidatorUpdate   `json:"validator_updates"`
+	ConsensusParamUpdates *cmtproto.ConsensusParams `json:"consensus_param_updates"`
+	AppHash               []byte                    `json:"app_hash"`
+}
+
+// BlockResultsServiceClient provides the block results of a given height (or latest if none provided).
+type BlockResultsServiceClient interface {
+	GetBlockResults(ctx context.Context, height *int64) (*BlockResults, error)
```
Comment (thanethomson) [inline rpc/grpc/client/block_results_service.go:26]: Would it not be better to have `height` be of type `int64`, similar to the block service's client interface?

Using `int64` seems more ergonomic than having to supply pointers to this function.

Comment (glnro) [reply rpc/grpc/client/block_results_service.go]: See [the latest commit](https://github.com/cometbft/cometbft/pull/1168/commits/14372cb1c560f77d96f4da3cd73cdacd86c359d3) with an update to provide an additional endpoint specifically to retrieve `BlockResults` of the latest committed height.

File: rpc/grpc/server/services/blockresultservice/service.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+// GetBlockResults returns the block results of the requested height.
+func (s *blockResultsService) GetBlockResults(_ context.Context, req *brs.GetBlockResultsRequest) (*brs.GetBlockResultsResponse, error) {
+	logger := s.logger.With("endpoint", "GetBlockResults")
+	ss, err := s.stateStore.Load()
+	if err != nil {
+		logger.Error("Error loading store", "err", err)
+		return nil, status.Error(codes.Internal, "Internal server error")
+	}
+	if req.Height > ss.LastBlockHeight || req.Height < 0 {
+		logger.Error("Error validating GetBlockResults request height")
```
Comment (thanethomson) [inline rpc/grpc/server/services/blockresultservice/service.go:40]:
```suggestion
```

Could pollute operators' logs. And the full error's returned to the caller anyways.

File: test/e2e/tests/grpc_test.go [hunk 17]
```diff
@@ -132,3 +135,60 @@ func TestGRPC_Block_GetLatestHeight(t *testing.T) {
...
+		require.GreaterOrEqual(t, last, latestBlockResults.Height)
+		require.NoError(t, err, "Unexpected error for GetLatestBlockResults")
+		require.NotNil(t, latestBlockResults)
+
+		successCases := []struct {
+			expectedHeight int64
+			requestHeight  int64
+		}{
+			{first, first},
+			{latestBlockResults.Height, latestBlockResults.Height},
```
Comment (thanethomson) [inline test/e2e/tests/grpc_test.go:173]: The expected and requested should always be the same, right? If so I suspect we only need `requestHeight` in the test cases.

File: test/e2e/tests/grpc_test.go [hunk 18]
```diff
@@ -132,3 +135,60 @@ func TestGRPC_Block_GetLatestHeight(t *testing.T) {
...
+		}
+		errorCases := []struct {
+			requestHeight int64
+		}{
+			{-1},
+			{100000},
```
Comment (thanethomson) [inline test/e2e/tests/grpc_test.go:179]: This may end up being non-deterministic. I'd recommend doing something like:

```suggestion
			{first-1},
			{last+10},
```

**Review by thanethomson — APPROVED** [review]
Looking good overall! Just a few minor recommendations, but then I'd say this is good to merge.

File: test/e2e/tests/grpc_test.go (hunk 17 above)
Comment (glnro) [reply test/e2e/tests/grpc_test.go:173]: Yes fair, this is leftover from before but it's not needed now

File: test/e2e/tests/grpc_test.go (hunk 18 above)
Comment (glnro) [reply test/e2e/tests/grpc_test.go]: See [updated test](https://github.com/cometbft/cometbft/pull/1168/commits/8d3a2a6f7f209a6cf749d4c4a37a31154cb0fb8f)

File: config/toml.go [hunk 19]
```diff
@@ -280,6 +280,11 @@ enabled = {{ .GRPC.VersionService.Enabled }}
 [grpc.block_service]
 enabled = {{ .GRPC.BlockService.Enabled }}
 
+# The gRPC block results service returns block results for a given height. If no height
+# is given, it will return the block results from the latest height.
```
Comment (jmalicevic) [inline config/toml.go:284]: This text might be before your separation of the services into `GetLatestHeight` and `GetHeight`. Atm in the code if the req.Height is 0 we do not return the latest height right?
